java.util.concurrent.TimeoutException: Idle timeout expired: 300015/300000 ms

Quoting:

> Is there anything we can do to prevent this from happening?

![image](https://user-images.githubusercontent.com/6345473/87705096-5f266000-c76b-11ea-9809-00cc3da1278e.png)


> Do you think it's possibly a problem if they had two browser tabs open on the same page and they start competing using FileIO?

```
2020-07-16 08:32:31,128 [ERROR] Failed to read '/Users/jack/Downloads/io/MYFILE.IN' (does not exist)
2020-07-16 08:33:02,130 [DEBUG] Message: {"call":"file.read","promise":{},"params":{"path":"/Users/jack/Downloads/io/MYFILE.IN","sandbox":false,"shared":false},"timestamp":1594913582009,"uid":"zbv44m","position":{"x":768,"y":491.5}}
2020-07-16 08:33:02,134 [ERROR] Failed to read '/Users/jack/Downloads/io/MYFILE.IN' (does not exist)
2020-07-16 08:33:33,126 [DEBUG] Message: {"call":"file.read","promise":{},"params":{"path":"/Users/jack/Downloads/io/MYFILE.IN","sandbox":false,"shared":false},"timestamp":1594913612996,"uid":"utizud","position":{"x":768,"y":491.5}}
2020-07-16 08:33:33,130 [ERROR] Failed to read '/Users/jack/Downloads/io/MYFILE.IN' (does not exist)
2020-07-16 08:34:04,087 [DEBUG] Message: {"call":"file.read","promise":{},"params":{"path":"/Users/jack/Downloads/io/MYFILE.IN","sandbox":false,"shared":false},"timestamp":1594913643954,"uid":"bid1s0","position":{"x":768,"y":491.5}}
2020-07-16 08:34:04,092 [ERROR] Failed to read '/Users/jack/Downloads/io/MYFILE.IN' (does not exist)
2020-07-16 08:34:06,084 [DEBUG] Message: {"call":"file.read","promise":{},"params":{"path":"/Users/jack/Downloads/mp2/AUTOMAT.IN","sandbox":false,"shared":false},"timestamp":1594913645968,"uid":"9pg4cy","position":{"x":768,"y":491.5}}
2020-07-16 08:39:06,095 [ERROR] Connection error
org.eclipse.jetty.websocket.api.CloseException: java.util.concurrent.TimeoutException: Idle timeout expired: 300000/300000 ms
at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onReadTimeout(AbstractWebSocketConnection.java:560)
at org.eclipse.jetty.io.AbstractConnection.onFillInterestedFailed(AbstractConnection.java:172)
at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillInterestedFailed(AbstractWebSocketConnection.java:533)
at org.eclipse.jetty.io.AbstractConnection$ReadCallback.failed(AbstractConnection.java:317)
at org.eclipse.jetty.io.FillInterest.onFail(FillInterest.java:138)
at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillableFail(SslConnection.java:466)
at org.eclipse.jetty.io.ssl.SslConnection.onFillInterestedFailed(SslConnection.java:329)
at org.eclipse.jetty.io.ssl.SslConnection$2.failed(SslConnection.java:164)
at org.eclipse.jetty.io.FillInterest.onFail(FillInterest.java:138)
at org.eclipse.jetty.io.AbstractEndPoint.onIdleExpired(AbstractEndPoint.java:407)
at org.eclipse.jetty.io.IdleTimeout.checkIdleTimeout(IdleTimeout.java:171)
at org.eclipse.jetty.io.IdleTimeout.idleCheck(IdleTimeout.java:113)
at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.util.concurrent.TimeoutException: Idle timeout expired: 300000/300000 ms
... 8 more
```

Quoting @bberenz:

> Well that just sounds like the tab with the connection is in the background

But @bberenz we have a ping/heartbeat preventing the websocket from every closing as idle, no?  Why would this ping/heartbeat stop?  is there something in FileIO that locks the websocket from sending a ping back?

@jkatsiotis thanks kindly for these steps.

I tested the same thing using QZ Tray 2.1.2 running on a Mac and when I sever the network connection on the 2nd computer, it immediately throws a close event `CloseEvent {isTrusted: true, wasClean: false, code: 1006, reason: "", type: "close", …}` to the browser.

Java does eventually throw an exception (about 5 minutes later), but this seems to be mostly benign.  I guess we could choose to suppress this error.  It doesn't prevent me from reconnecting like your steps illustrate.

```
	Connection error
org.eclipse.jetty.websocket.api.CloseException: java.util.concurrent.TimeoutException: Idle timeout expired: 300070/300000 ms
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onReadTimeout(AbstractWebSocketConnection.java:560)
	at org.eclipse.jetty.io.AbstractConnection.onFillInterestedFailed(AbstractConnection.java:172)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillInterestedFailed(AbstractWebSocketConnection.java:533)
	[... truncated ...]
Caused by: java.util.concurrent.TimeoutException: Idle timeout expired: 300070/300000 ms
	... 8 more
[INFO] 2021-02-04 14:23:51,497 @ qz.ws.PrintSocketClient:?
	Connection closed: 1001 - java.util.concurrent.TimeoutException: Idle timeout expired: 300070/300000 ms
```

Are you saying you can't reconnect immediately after the network recovers?

According to this article, the WebSocket isn't intended to reconnect by design, however I'm curious why the behavior is different.
https://stackoverflow.com/a/13797337/3196753

@bberenz do you think the timeout error should be suppressed from the user?

Oddly, I was able to reproduce this by starting a second instance of QZ Tray on the same machine as a non-administrator account... here's the logs... 

```
2022-06-28 11:52:57,542 [DEBUG] Checking for a running instance of QZ Tray on port 8182
2022-06-28 11:52:59,060 [DEBUG] Checking for a running instance of QZ Tray on port 8384
2022-06-28 11:53:01,430 [DEBUG] Checking for a running instance of QZ Tray on port 8485
2022-06-28 11:53:02,248 [WARN] WebSocket error
org.eclipse.jetty.websocket.api.UpgradeException: 0 null
	at org.eclipse.jetty.websocket.client.WebSocketUpgradeRequest.onComplete(WebSocketUpgradeRequest.java:524)
	at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:198)
	at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:190)
	at org.eclipse.jetty.client.HttpReceiver.terminateResponse(HttpReceiver.java:444)
	at org.eclipse.jetty.client.HttpReceiver.abort(HttpReceiver.java:527)
	at org.eclipse.jetty.client.HttpChannel.abortResponse(HttpChannel.java:156)
	at org.eclipse.jetty.client.HttpChannel.abort(HttpChannel.java:149)
	at org.eclipse.jetty.client.HttpExchange.abort(HttpExchange.java:257)
	at org.eclipse.jetty.client.HttpConversation.abort(HttpConversation.java:149)
	at org.eclipse.jetty.client.HttpRequest.abort(HttpRequest.java:768)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.abort(HttpConnectionOverHTTP.java:193)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.close(HttpConnectionOverHTTP.java:179)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onIdleExpired(HttpConnectionOverHTTP.java:144)
	at org.eclipse.jetty.io.AbstractEndPoint.onIdleExpired(AbstractEndPoint.java:402)
	at org.eclipse.jetty.io.IdleTimeout.checkIdleTimeout(IdleTimeout.java:171)
	at org.eclipse.jetty.io.IdleTimeout.idleCheck(IdleTimeout.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.concurrent.TimeoutException: Idle timeout 3000 ms
	... 10 more
2022-06-28 11:53:05,220 [INFO] Server started on port(s) 8283, 8282
2022-06-28 11:53:44,923 [INFO] SSL certificate is still valid for 824 more days: Sun Sep 29 11:44:45 EDT 2024.  We'll make a new one automatically when needed.
```

--- a/src/qz/ws/PrintSocketClient.java
+++ b/src/qz/ws/PrintSocketClient.java
@@ -10,8 +10,8 @@ import org.codehaus.jettison.json.JSONObject;
 import org.eclipse.jetty.server.Server;
 import org.eclipse.jetty.websocket.api.Session;
 import org.eclipse.jetty.websocket.api.annotations.*;
-import org.eclipse.jetty.websocket.api.exceptions.CloseException;
 import org.eclipse.jetty.websocket.api.exceptions.WebSocketException;
+import org.eclipse.jetty.websocket.api.exceptions.WebSocketTimeoutException;
 import org.eclipse.jetty.websocket.server.JettyServerUpgradeRequest;
 import org.eclipse.jetty.websocket.server.JettyServerUpgradeResponse;
 import org.usb4java.LoaderException;
@@ -39,7 +39,6 @@ import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Locale;
 import java.util.concurrent.Semaphore;
-import java.util.concurrent.TimeoutException;
 
 
 @WebSocket
@@ -90,7 +89,7 @@ public class PrintSocketClient {
     public void onError(Session session, Throwable error) {
         if (error instanceof EOFException || error instanceof ClosedChannelException) { return; }
 
-        if (error instanceof CloseException && error.getCause() instanceof TimeoutException) {
+        if (error instanceof WebSocketTimeoutException) {
             log.error("Timeout error (Lost connection with client)", error);
             return;
         }

